src/sentry/search/django/backend.py

```diff
@@ -8,275 +8,397 @@
+
+        group_queryset = QuerySetBuilder({
+            'query': CallbackCondition(
+                lambda queryset, query: queryset.filter(
+                    Q(message__icontains=query) | Q(culprit__icontains=query),
+                ) if query else queryset,
+            ),
+            'status': CallbackCondition(
+                lambda queryset, status: queryset.filter(status=status),
+            ),
+            'bookmarked_by': CallbackCondition(
+                lambda queryset, user: queryset.filter(
+                    bookmark_set__project=project,
+                    bookmark_set__user=user,
+                ),
+            ),
+            'assigned_to': CallbackCondition(
+                lambda queryset, user: queryset.filter(
+                    assignee_set__project=project,
+                    assignee_set__user=user,
+                ),
+            ),
+            'unassigned': CallbackCondition(
+                lambda queryset, unassigned: queryset.filter(
+                    assignee_set__isnull=unassigned,
+                ),
+            ),
+            'subscribed_by': CallbackCondition(
+                lambda queryset, user: queryset.filter(
+                    id__in=GroupSubscription.objects.filter(
+                        project=project,
+                        user=user,
+                        is_active=True,
+                    ).values_list('group'),
+                ),
+            ),
+            'active_at_from': ScalarCondition('active_at', 'gt'),
+            'active_at_to': ScalarCondition('active_at', 'lt'),
+        }).build(
+            Group.objects.filter(project=project).exclude(status__in=[
+                GroupStatus.PENDING_DELETION,
+                GroupStatus.DELETION_IN_PROGRESS,
+                GroupStatus.PENDING_MERGE,
+            ]),
+            parameters,
+        )
 
         if environment is not None:
-            # XXX: This overwrites the ``environment`` tag, if present, to
-            # ensure that the result set is limited to groups that have been
-            # seen in this environment (there is no way to search for groups
-            # that match multiple values of a single tag without changes to the
-            # tagstore API.)
-            tags['environment'] = environment.name
-
-        if tags:
-            matches = tagstore.get_group_ids_for_search_filter(
-                project.id,
-                environment.id if environment is not None else None,
-                tags,
-            )
-            if not matches:
-                return queryset.none()
-            queryset = queryset.filter(
-                id__in=matches,
+            if 'environment' in tags:
+                # TODO: This should probably just overwrite the existing tag,
+                # rather than asserting on it, but...?
+                assert Environment.objects.get(
+                    projects=project,
+                    name=tags.pop('environment'),
```

Comment:
This is removed since it's implied with the join onto `GroupEnvironment` below on line 296 and would otherwise cause an unneeded pass in the tagstore filters.